src/v3/core/graph.js

```diff
@@ -20,22 +21,72 @@ export type NeighborsOptions = {|
 
 export opaque type GraphJSON = any; // TODO
 
+const BASE_NODE_ADDRESS = "N\0";
+const BASE_EDGE_ADDRESS = "E\0";
+const SEPARATOR = "\0";
+
+function assertNodeAddress(x: NodeAddress) {
+  if (x == null) {
+    throw new Error(String(x));
+  }
+  if (!x.startsWith(BASE_NODE_ADDRESS)) {
+    if (x.startsWith(BASE_EDGE_ADDRESS)) {
+      throw new Error(`Expected NodeAddress, got EdgeAddress: ${x}`);
+    }
+    throw new Error(`Malformed NodeAddress: ${x}`);
+  }
+}
+
+function assertEdgeAddress(x: EdgeAddress) {
+  if (x == null) {
+    throw new Error(String(x));
+  }
+  if (!x.startsWith(BASE_EDGE_ADDRESS)) {
+    if (x.startsWith(BASE_NODE_ADDRESS)) {
+      throw new Error(`Expected EdgeAddress, got NodeAddress: ${x}`);
+    }
+    throw new Error(`Malformed EdgeAddress: ${x}`);
+  }
+}
+
+function assertAddressArray(arr: $ReadOnlyArray<string>) {
+  if (arr == null) {
+    throw new Error(String(arr));
+  }
+  arr.forEach((s: string) => {
+    const nullIdx = s.indexOf(SEPARATOR);
+    if (nullIdx !== -1) {
+      throw new Error(`NUL char: ${stringify(arr)}`);
+    }
+  });
+}
+
 export function toNodeAddress(arr: $ReadOnlyArray<string>): NodeAddress {
-  const _ = arr;
-  throw new Error("toNodeAddress");
+  assertAddressArray(arr);
+  if (arr.length > 0) {
```

Comment:
Okay. This implementation isn’t quite as neat as I’d like (something
like `["N", ...arr, ""].join(SEPARATOR)` would be better, in my
opinion), but not a big deal.